aws-lc-rs/src/rsa/oaep.rs

```diff
@@ -0,0 +1,472 @@
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs during deserialization of this key from PKCS#8.
+    pub fn from_pkcs8(pkcs8: &[u8]) -> Result<Self, KeyRejected> {
+        unsafe {
+            let evp_pkey = LcPtr::try_from(pkcs8)?;
+            super::key::validate_rsa_pkey(&evp_pkey)?;
+            Self::new(evp_pkey).map_err(|_| KeyRejected::unexpected_error())
+        }
+    }
+
+    /// Returns the corresponding [`KeySize`].
+    #[must_use]
+    pub fn key_size(&self) -> KeySize {
+        self.size
+    }
+
+    /// Retrieves the `PublicEncryptingKey` corresponding with this `PrivateDecryptingKey`.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs computing the public key.
+    pub fn public_key(&self) -> Result<PublicEncryptingKey, Unspecified> {
+        if 1 != unsafe { EVP_PKEY_up_ref(*self.key) } {
+            return Err(Unspecified);
+        };
+        PublicEncryptingKey::new(LcPtr::new(*self.key)?)
+    }
+
+    /// Decrypts the contents in `ciphertext` and writes the corresponding plaintext to `output`.
+    ///
+    /// # Errors
+    /// * `Unspeicifed` for any error that occurs while decrypting `ciphertext`.
+    pub fn decrypt<'output>(
+        &self,
+        algorithm: &'static EncryptionAlgorithm,
+        ciphertext: &[u8],
+        output: &'output mut [u8],
+    ) -> Result<&'output mut [u8], Unspecified> {
+        let pkey_ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.key, null_mut()) })?;
+
+        if 1 != unsafe { EVP_PKEY_decrypt_init(*pkey_ctx) } {
+            return Err(Unspecified);
+        }
+
+        configure_oaep_crypto_operation(
+            &pkey_ctx,
+            algorithm.oaep_hash_fn(),
+            algorithm.mgf1_hash_fn(),
+        )?;
+
+        let mut out_len = output.len();
+
+        if 1 != indicator_check!(unsafe {
+            EVP_PKEY_decrypt(
+                *pkey_ctx,
+                output.as_mut_ptr(),
+                &mut out_len,
+                ciphertext.as_ptr(),
+                ciphertext.len(),
+            )
+        }) {
+            return Err(Unspecified);
+        };
+
+        Ok(&mut output[..out_len])
+    }
+}
+
+impl AsDer<Pkcs8V1Der<'static>> for PrivateDecryptingKey {
+    fn as_der(&self) -> Result<Pkcs8V1Der<'static>, Unspecified> {
+        let mut buffer = [0u8; PKCS8_FIXED_CAPACITY_BUFFER];
+        let mut cbb = LcCBB::new_fixed(&mut buffer);
+
+        if 1 != unsafe { EVP_marshal_private_key(cbb.as_mut_ptr(), *self.key.as_const()) } {
+            return Err(Unspecified);
+        }
+        let out_len = cbb.finish()?;
+
+        Ok(Buffer::take_from_slice(&mut buffer[..out_len]))
+    }
+}
+
+/// An RSA Public Key used for decrypting ciphertext encrypted by [`PublicEncryptingKey`].
+pub struct PublicEncryptingKey {
+    key: LcPtr<EVP_PKEY>,
+    size: KeySize,
+}
+
+impl PublicEncryptingKey {
+    fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
+        if !is_rsa_evp_pkey(&key) {
```

Comment:
* This check is called both here and in `KeySize::from_evp_pkey`.  
* To avoid some of the redundancy between `PublicEncryptingKey` and `PrivateDecryptingKey` you could make them wrappers around a common type: 
```rust
pub struct PublicEncryptingKey(RsaEvpPkey);

impl PublicEncryptingKey {
    fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
        Ok(Self(RsaEvpPkey::new(key)?))
    }
...
}

pub struct PrivateDecryptingKey(RsaEvpPkey)

impl PrivateDecryptingKey {
    fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
        Ok(Self(RsaEvpPkey::new(key)?))
    }
...
}


pub(crate) struct RsaEvpPkey {
    key: LcPtr<EVP_PKEY>,
    size: KeySize,
}
 impl RsaEvpPkey {
     pub(crate) fn new(key: LcPtr<EVP_PKEY>) -> Result<Self, Unspecified> {
        if EVP_PKEY_RSA != unsafe { EVP_PKEY_id(key.as_const_ptr()) } {
            return Err(Unspecified);
        }
        let size = match unsafe { EVP_PKEY_size(key.as_const_ptr()) } {
            256 => Ok(KeySize::Rsa2048),
            512 => Ok(KeySize::Rsa4096),
            1024 => Ok(KeySize::Rsa8192),
            _ => Err(Unspecified),
        }?;
        Ok(Self { key, size })
    }
...
}
```